R/correlation.R

```diff
@@ -340,66 +348,135 @@ Correlation <- function(jaspResults, dataset, options){
                                   vvars      = vcomb[[i]],
                                   res        = currentResults,
-                                  errors     = errors, 
+                                  errors     = errors,
                                   testErrors = testErrors,
                                   shapiro    = shapiro)
-      
+
       # store state for pair
       state <- createJaspState(object = results[[vpair[i]]])
       state$dependOn(options = c("conditioningVariables", "hypothesis",
                                  "confidenceIntervals", "confidenceIntervalsInterval",
                                  "missingValues"),
                      optionContainsValue = list(variables = .unv(vcomb[[i]])))
-      
+
       jaspResults[[vpair[i]]] <- state
     }
-    
+
     #progressbarTick()
   }
-  
-  
+
+  if(options[["bootstrap"]])
+    results <- .corrCalculateBootstrapCI(results, dataset, options)
+
   jaspResults[['results']] <- createJaspState(object = results)
   jaspResults[['results']]$dependOn(options = c("variables", "conditioningVariables",  "hypothesis",
                                                 "confidenceIntervalsInterval", "missingValues",
-                                                "pearson", "spearman", "kendallsTauB"))
-  
-  
+                                                "pearson", "spearman", "kendallsTauB",
+                                                "bootstrap", "bootstrapReplicates"))
+
+
   return(results)
 }
 
+.corrCalculateBootstrapCI <- function(results, dataset, options) {
+  alpha <- options[["confidenceIntervalsInterval"]]
+  ciLevel <- c((1-alpha)/2, 1-(1-alpha)/2)
+
+  # bootstrapping function (returns bootstraps per variable pair and test)
+  .corrBoot <- function() {
+    idx <- sample(x = 1:nrow(dataset), size = nrow(dataset), replace = TRUE)
+    data <- dataset[idx,]
+
+    if(length(options[["conditioningVariables"]]) > 0) {
+      z <- data[,options[["conditioningVariables"]], drop=FALSE]
+    } else {
+      z <- NULL
+    }
+
+    out <- list()
+    for(name in names(results)) {
+      out[[name]] <- list()
+      xy <- data[, results[[name]]$vars]
+
+      for(method in c('pearson', 'spearman', 'kendall')){
+        compute <- .corrTestChecked(method, options)
+        if(compute) {
+          out[[name]][[method]] <- try(.corrFastParCor(xy = xy, z = z, method = method))
+        } else {
+          out[[name]][[method]] <- NaN
+        }
+
+        if(isTryError(out[[name]][[method]])) out[[name]][[method]] <- NA
+      }
+    }
+
+    progressbarTick()
+    return(out)
+  }
+
+  # bootstrap and reshape
+  startProgressbar(expectedTicks = options[["bootstrapReplicates"]], label = gettext("Bootstrapping"))
+  bootstraps <- replicate(options[["bootstrapReplicates"]], .corrBoot(), simplify = FALSE)
+  bootstraps <- purrr::transpose(bootstraps)
+  bootstraps <- lapply(bootstraps, function(b) do.call(rbind, b))
```

Comment:
So I see what you're doing here, but I don't really like it. First, you create a list with shape:

- Bootstrap 1
  - contNormal_contGamma
    - Pearson
      - value
    - Kendall
      - value
    - Spearman
      - value
  - contNormal_contcor1
    - Pearson
      - value
    - Kendall
      - value
    - Spearman
      - value
- Bootstrap 2
  - ...

Next, this structure turns out to be not very practical so you do 
```r
bootstraps <- replicate(options[["bootstrapReplicates"]], .corrBoot(), simplify = FALSE)
bootstraps <- purrr::transpose(bootstraps)
bootstraps <- lapply(bootstraps, function(b) do.call(rbind, b))
```
to simplify it. The resulting type of `bootstrap` is a list with 3 matrices. The elements of these matrices are actually lists (which is why you need the `unlist` when computing the quantiles). 

The downside here is that lists take up more memory and are slower for subsetting when you want to say lookup all values in a column (like when computing the quantiles).


My suggestion would be to create a 3D array, like so:
```r
bootstraps <- array(NaN, dim = c(options[["bootstrapReplicates"]], length(methods), length(results)), 
                    dimnames = list(NULL, methods, names(results)))
```
and then loop explicitly over the bootstraps:
```r
# do these two things once outside of the loop
methodsForLoop <- methods[which(vapply(methods, function(method) .corrTestChecked(method, options), logical(1L)))]
if(length(options[["conditioningVariables"]]) > 0) {
  z <- data[,options[["conditioningVariables"]], drop=FALSE]
} else {
  z <- NULL
}
for (i in seq_len(options[["bootstrapReplicates"]])) {

    # slightly faster than `sample(x = 1:nrow(dataset), size = nrow(dataset), replace = TRUE)`
    idx <- sample(x = nrow(dataset), replace = TRUE)
    data <- dataset[idx,]

    for(name in names(results)) {
      xy <- data[, results[[name]]$vars]
      
      for(method in methodsForLoop) {
        
        # Since we initialized with NaN, this becomes a bit simpler
        bootstraps[i, method, name] <- tryCatch({
          .corrFastParCor(xy = xy, z = z, method = method)
        }, error = function() NaN
        )
      }
    }

    progressbarTick()

  }
```
Note that I didn't test this code, but you get the idea.

Reply:
Busted! ☹️ Yeah I'll do it your way then